Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: MagicNumber

child of #14942 :

doc :https://checkstyle.org/checks/coding/magicnumber.html


---

```
PS D:\CS\test> cat src/Test.java  
import java.util.LinkedHashMap;
import java.util.List;

record ColoredPoint(String p, String x, int c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {
       int _ = 4;

    }
}

PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="MagicNumber">
        </module>
    </module>
</module>
PS D:\CS\test> javac src/Test.java
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:16: '4' is a magic number. [MagicNumber]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 

```

---

The check always detects magic numbers whether it is assigned to an unnamed variable or not.

This check is fine on patterns with a `when` guard:
```
➜  src javac Test.java
➜  src cat Test.java  
public class Test {
    public static void main(String[] args) {

       Object o = new Object();

        if (o instanceof Integer i && i < 3) { // violation
            System.out.println("null");
        } else {
            System.out.println("not null");
        }

        switch (o) {
            case Integer i when i < 3: // violation
                System.out.println("null");
                break;
            default:
                System.out.println("not null");
        }
    }
}
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="MagicNumber">
        </module>
    </module>
</module>
➜  src java -jar checkstyle-10.17.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:6:43: '3' is a magic number. [MagicNumber]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:13:37: '3' is a magic number. [MagicNumber]
Audit done.
Checkstyle ends with 2 errors.
```

We can skip opening an issue for this if we don't already have one.

This check will not apply to unnamed patterns at all, since `_` can never be referenced (therefore we cannot use it in a `when` guard where we would have some conditional).

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheckTest.java
@@ -744,4 +744,16 @@ public class MagicNumberCheckTest
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputMagicNumberMagicNumberInGuards.java"), expected);
     }
+
+    @Test
+    public void testMagicNumberWithUnnamedVariables() throws Exception {
+        final String[] expected = {
+            "20:17: " + getCheckMessage(MSG_KEY, "9"),
+            "22:21: " + getCheckMessage(MSG_KEY, "17"),
+            "23:20: " + getCheckMessage(MSG_KEY, "3.1415"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputMagicNumberMagicNumberWithUnnamedVariables.java"),
+                expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberMagicNumberWithUnnamedVariables.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberMagicNumberWithUnnamedVariables.java
@@ -1,0 +1,25 @@
+/*
+MagicNumber
+ignoreNumbers = (default)-1, 0, 1, 2
+ignoreHashCodeMethod = (default)false
+ignoreAnnotation = (default)false
+ignoreFieldDeclaration = (default)false
+ignoreAnnotationElementDefaults = (default)true
+constantWaiverParentToken = (default)TYPECAST, METHOD_CALL, EXPR, ARRAY_INIT, UNARY_MINUS, \
+                            UNARY_PLUS, ELIST, STAR, ASSIGN, PLUS, MINUS, DIV, LITERAL_NEW
+tokens = (default)NUM_DOUBLE, NUM_FLOAT, NUM_INT, NUM_LONG
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.magicnumber;
+
+public class InputMagicNumberMagicNumberWithUnnamedVariables {
+    void test() {
+        int _ = 9; // violation
+        int _ = 1;
+        Integer _ = 17; // violation
+        double _ = 3.1415; // violation
+    }
+}

